---
title: "Differential expression between celltypes"
subtitle: "Wilcox test (Seurat)"
author: "Paul Heinrich"
output: 
    html_document:
        toc: true
---

```{r setup1, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.dim = c(10, 10))
```

# Notes

Reference single cell dataset from "A single-cell survey of the small intestinal epithelium"
(Haber et al., 2017, Nature), available at NCBI GEO (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92332).

```{r setup2,echo=FALSE, warning=FALSE, message=FALSE}
#rbioc_3-14
library(rprojroot)
library(yaml)
library(knitr)

saveData <- TRUE

root_criterion <- ".projroot"
proot <- rprojroot::has_file(root_criterion)

source(proot$find_file("experiments/scRNA/analysis/custom/functions/custom_functions.R"))

params <- list()
params$config <- proot$find_file("experiments/scRNA/analysis/custom/hep62703_2023-03-24_A_harmony/config.yaml")
params$analparams <- proot$find_file("experiments/scRNA/analysis/custom/hep62703_2023-03-24_A_harmony/analparams_preprocessing_filters_2023-03-24.yaml")
params$date <- Sys.Date()

config <- yaml.load_file(proot$find_file(params$config))
analparams <- yaml.load_file(proot$find_file(params$analparams))
today <- params$date

print(today)

# libraries

libload <- c(
    "Seurat", "ggplot2", "harmony", "Nebulosa", "future",
    "WriteXLS", "knitr", "stringr", "dplyr", "tidyr"
)

for(i in libload) {
    library(i, character.only = T)
}

# Parallelization with future

plan("multicore", workers = 6)
plan()

set.seed(7)

```

```{r loaddata,eval=T,echo=F}

identCol <- "annot_SingleR_mod_monocle_Stem.TA_ref_regress.cc"

SObj <- readRDS(proot$find_file(analparams$Seurat_clustered))
Idents(SObj) <- SObj[[identCol]]

```

```{r diff_expr, eval=T, echo=F}

comparisons <- list(
    c("Stem.TA", "Stem"),
    c("TA", "Stem.TA"),
    c("TA", "Stem")
)
names(comparisons) <- sapply(comparisons, function(x) paste(x, collapse = "_vs_"))

diffExpr <- list()

for (comparison in names(comparisons)) {
    print(Sys.time())
    print(paste0("Comparison: ", comparison))

    diffExpr[[comparison]] <- FindMarkers(
        SObj,
        ident.1 = comparisons[[comparison]][1],
        ident.2 = comparisons[[comparison]][2],
        test.use = "wilcox",
        only.pos = FALSE,
        min.pct = 0.01,
        logfc.threshold = -Inf,
        return.thresh = 1,
    )
    diffExpr[[comparison]]$p_val_adj_fdr <- p.adjust(diffExpr[[comparison]]$p_val, method = "fdr")
}

fractionBelowThreshold <- c()
threshold <- 0.05

for (comparison in names(comparisons)) {
    fractionBelowThreshold[comparison] <- sum(diffExpr[[comparison]]$p_val <= threshold)/
    length(diffExpr[[comparison]]$p_val)
}

print(fractionBelowThreshold)

```

```{r save_data, echo=FALSE, warning=FALSE, eval=saveData}

commonFilename <- paste0("diff_expr_Stem-TA_wilcox_monocle_Stem-TA_ref_regress-cc_", today)

saveRDS(
    diffExpr,
    proot$find_file(config$outputpath, "data", paste0(commonFilename, ".rds"))
)

WriteXLS(
    lapply(diffExpr, function(x) tibble::rownames_to_column(x, "gene")),
    proot$find_file(config$outputpath, "table", paste0(commonFilename, ".xlsx")),
)

```

# SessionInfo

```{r sessionInfo}

sessionInfo()

```

